includes/class-wcs-action-scheduler-customer-notifications.php

```diff
@@ -0,0 +1,290 @@
+	}
+
+	public function set_hours_offset( $hours_offset ) {
+		$this->hours_offset = $hours_offset;
+	}
+
+	public function __construct() {
+		parent::__construct();
+
+		$setting_option     = get_option(
+			WC_Subscriptions_Admin::$option_prefix . WC_Subscriptions_Email_Notifications::$offset_setting_string,
+			array(
+				'number' => 3,
+				'unit'   => 'days',
+			)
+		);
+		$this->hours_offset = self::convert_offset_to_hours( $setting_option );
+	}
+
+	/**
+	 * Calculate time offset in hours from the settings array.
+	 *
+	 * @param array $offset Format: array( 'number' => 3, 'unit' => 'days' )
+	 *
+	 * @return float|int
+	 */
+	protected static function convert_offset_to_hours( $offset ) {
+		$default_offset = 3 * DAY_IN_SECONDS / HOUR_IN_SECONDS;
+
+		if ( ! isset( $offset['unit'] ) || ! isset( $offset['number'] ) ) {
+			return $default_offset;
+		}
+
+		switch ( $offset['unit'] ) {
+			case 'days':
+				return ( $offset['number'] * DAY_IN_SECONDS / HOUR_IN_SECONDS );
+			case 'weeks':
+				return ( $offset['number'] * WEEK_IN_SECONDS / HOUR_IN_SECONDS );
+			case 'months':
+				return ( $offset['number'] * MONTH_IN_SECONDS / HOUR_IN_SECONDS );
+			case 'years':
+				return ( $offset['number'] * YEAR_IN_SECONDS / HOUR_IN_SECONDS );
+			default:
+				return $default_offset;
+		}
+	}
+
+	public function set_date_types_to_schedule() {
+		$date_types_to_schedule = wcs_get_subscription_date_types();
+		unset(
+			$date_types_to_schedule['start'],
+			$date_types_to_schedule['cancelled'] // prevent scheduling end date when reactivating subscription.
+		);
+
+		//TODO: filter?
+		$this->date_types_to_schedule = array_keys( $date_types_to_schedule );
+	}
+
+	protected function schedule_notification( $subscription, $action, $timestamp ) {
+		if (
+			! (
+				$subscription->has_status( 'active' )
+				|| $subscription->has_status( 'pending-cancel' ) //TODO: do we want to create notifications when user cancelled the subscription?
+			)
+		) {
+			return;
+		}
+
+		$action_args = $this->get_action_args( $subscription );
+
+		$next_scheduled = as_next_scheduled_action( $action, $action_args );
+
+		if ( $timestamp === $next_scheduled ) {
+			return;
+		}
+
+		$this->unschedule_actions( $action, $action_args );
+
+		// Only reschedule if it's in the future
+		if ( $timestamp <= time() ) {
+			return;
+		}
+
+		as_schedule_single_action( $timestamp, $action, $action_args );
+	}
+
+	//TODO: check timezones
+	/*
+	 * Subtract time offset from given datetime based on the settings and subscription properties.
+	 *
+	 * @param string $datetime
+	 * @param WC_Subscription $subscription
+	 *
+	 * @return int
+	 */
+	protected function sub_time_offset( $datetime, $subscription ) {
+		$dt = new DateTime( $datetime, new DateTimeZone( 'UTC' ) );
+		$dt->sub( new DateInterval( "PT{$this->get_hours_offset( $subscription )}H" ) );
+
+		return $dt->getTimestamp();
+	}
```

Comment:
I might be missing the rest of the context here but I'm wondering if we should just use the number of seconds and timestamps? 

I noticed that the setting (interval and period) is being converted into hours and I'm wondering if it would be easier to convert it into seconds and then just do a straight subtraction.

Reply:
Seems like a good idea to me. Working with seconds allows us more granular control. 
Also, since both the subscription's `get_date()` and the AS work in UTC by default, we should be aligned. Right?

Reply:
I generally prefer to work with standard DateTime classes when available to avoid some of the pitfalls mentioned here https://gist.github.com/timvisee/fcda9bbdff88d45cc9061606b4b923ca, but I suppose using timestamps would be easier here and should be mostly ok.

Reply:
Switched to timestamps, so resolving this.